### PR TITLE
Added organize feature

### PR DESCRIPTION
**Pull request by deon-gracias**
From issue [61](https://github.com/unidoc/unipdf-cli/issues/61)

Adds the ability to organize a pdf's pages.

image

### CODE REVIEW

File: internal/cli/utils.go [hunk 1]
```diff
@@ -86,6 +86,62 @@ func parsePageRange(pageRange string) ([]int, error) {
...
+	var pages []int
+
+	rngs := strings.Split(removeSpaces(pageRange), ",")
+	for _, rng := range rngs {
+		if rng == "" {
+			continue
+		}
+
+		indices := strings.Split(rng, "-")
+
```
Comment (3ace) [inline internal/cli/utils.go:99]: could you remove this empty line?

File: internal/cli/utils.go [hunk 2]
```diff
@@ -86,6 +86,62 @@ func parsePageRange(pageRange string) ([]int, error) {
...
+		lenIndices := len(indices)
+		if lenIndices > 2 {
+			return nil, errors.New("invalid page range")
+		}
+		if lenIndices == 2 {
+			start, err := strconv.Atoi(indices[0])
+			if err != nil {
+				return nil, errors.New("invalid page number")
```
Comment (3ace) [inline internal/cli/utils.go:107]: it might be better to rephrase this as
```suggestion
				return nil, errors.New("invalid start page number")
```

File: internal/cli/utils.go [hunk 3]
```diff
@@ -86,6 +86,62 @@ func parsePageRange(pageRange string) ([]int, error) {
...
+			}
+			if start < 1 {
+				return nil, errors.New("page range start must be greater than 0")
+			}
+
+			end, err := strconv.Atoi(indices[1])
+			if err != nil {
+				return nil, errors.New("invalid page number")
```
Comment (3ace) [inline internal/cli/utils.go:115]: and this into
```suggestion
				return nil, errors.New("invalid end page number")
```

File: internal/cli/organize.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		}
+
+		if err := pdf.Organize(inputPath, outputPath, password, pages); err != nil {
+			printErr("Error: %s\n", err)
+		}
+
+		fmt.Printf("Successfully organized file %s\n", inputPath)
+		fmt.Printf("Output file saved to %s\n", outputPath)
+	},
+	Args: func(cmd *cobra.Command, args []string) error {
```
Comment (3ace) [inline internal/cli/organize.go:62]: Minor adjustment here to satisfy the linter

```suggestion
	Args: func(_ *cobra.Command, args []string) error {
```

File: pkg/pdf/organize.go
Comment (3ace) [inline pkg/pdf/organize.go:1]: could you replace the content of this file here with

```golang
/*
 * This file is subject to the terms and conditions defined in
 * file 'LICENSE.md', which is part of this source code package.
 */

package pdf

import (
	"github.com/unidoc/unipdf/v3/common"
	unipdf "github.com/unidoc/unipdf/v3/model"
)

// Organize extracts the provided page list from PDF file specified by the
// inputPath parameter then merges the individual pages and saves the
// resulting file at the location specified by the outputPath parameter.
// A password can be passed in for encrypted input files.
func Organize(inputPath, outputPath, password string, pages []int) error {
	// Read input file.
	pdfReader, _, _, _, err := readPDF(inputPath, password)
	if err != nil {
		return err
	}

	// Add selected pages to the writer.
	pdfWriter := unipdf.NewPdfWriter()

	for i := 0; i < len(pages); i++ {
		page, err := pdfReader.GetPage(pages[i])
		if err != nil {
			return err
		}

		err = pdfWriter.AddPage(page)
		if err != nil {
			return err
		}
	}

	// Copy PDF version.
	version := pdfReader.PdfVersion()
	pdfWriter.SetVersion(version.Major, version.Minor)

	// Copy PDF info.
	info, err := pdfReader.GetPdfInfo()
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
	} else {
		pdfWriter.SetDocInfo(info)
	}

	// Copy Catalog Metadata.
	if meta, ok := pdfReader.GetCatalogMetadata(); ok {
		if err := pdfWriter.SetCatalogMetadata(meta); err != nil {
			return err
		}
	}

	// Copy catalog mark information.
	if markInfo, ok := pdfReader.GetCatalogMarkInfo(); ok {
		if err := pdfWriter.SetCatalogMarkInfo(markInfo); err != nil {
			return err
		}
	}

	// Copy AcroForm.
	err = pdfWriter.SetForms(pdfReader.AcroForm)
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
		return err
	}

	// Copy viewer preferences.
	if pref, ok := pdfReader.GetCatalogViewerPreferences(); ok {
		if err := pdfWriter.SetCatalogViewerPreferences(pref); err != nil {
			return err
		}
	}

	// Copy language preferences.
	if lang, ok := pdfReader.GetCatalogLanguage(); ok {
		if err := pdfWriter.SetCatalogLanguage(lang); err != nil {
			return err
		}
	}

	// Copy document outlines.
	pdfWriter.AddOutlineTree(pdfReader.GetOutlineTree())

	// Copy OC Properties.
	props, err := pdfReader.GetOCProperties()
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
	} else {
		err = pdfWriter.SetOCProperties(props)
		if err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

	// Copy page labels.
	labelObj, err := pdfReader.GetPageLabels()
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
	} else {
		err = pdfWriter.SetPageLabels(labelObj)
		if err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

	// Copy named destinations.
	namedDest, err := pdfReader.GetNamedDestinations()
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
	} else {
		err = pdfWriter.SetNamedDestinations(namedDest)
		if err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

	// Copy name dictionary.
	nameDict, err := pdfReader.GetNameDictionary()
	if err != nil {
		common.Log.Debug("ERROR: %v", err)
	} else {
		err = pdfWriter.SetNameDictionary(nameDict)
		if err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

        // Copy StructTreeRoot dictionary.
	structTreeRoot, found := pdfReader.GetCatalogStructTreeRoot()
	if found {
		err := pdfWriter.SetCatalogStructTreeRoot(structTreeRoot)
		if err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

	// Copy global page rotation.
	if pdfReader.Rotate != nil {
		if err := pdfWriter.SetRotation(*pdfReader.Rotate); err != nil {
			common.Log.Debug("ERROR: %v", err)
		}
	}

	// Write output file.
	safe := inputPath == outputPath
	return writePDF(outputPath, &pdfWriter, safe)
}
```

This adds some calls to copy the original PDF file information to the new file and rename some variable names

**Review by 3ace — CHANGES_REQUESTED** [review]

File: internal/cli/utils.go (hunk 1 above)
Comment (deon-gracias) [reply internal/cli/utils.go:99]: This function is the exact copy of the `parsePageRange` function along with its formatting. Should I remove from both functions or  only the line 99 one?

File: pkg/pdf/organize.go
Comment (deon-gracias) [reply pkg/pdf/organize.go:1]: Changes added at on commit [d61c5e3a3fa371ce42d05efa2f860bf528381d52](https://github.com/deon-gracias/unipdf-cli/commit/d61c5e3a3fa371ce42d05efa2f860bf528381d52)

File: internal/cli/utils.go (hunk 1 above)
Comment (3ace) [reply internal/cli/utils.go:99]: sure, you could remove both of them.

Comment (deon-gracias) [reply internal/cli/utils.go]: Done :+1:

**Review by 3ace — APPROVED** [review]

**Review by sampila — APPROVED** [review]
Looks good to me
